### DISCUSSION TITLE
Creating a local copy of a PV

### DISCUSSION THREAD

**Discussion by JonFijalkowski** [ControlSystemStudio/phoebus#3213] [Q&A]
One of the main parts of the work I've been planning to do for https://github.com/ControlSystemStudio/phoebus/issues/3167 is reliant on copying a Live PV to some type of local PV/datastructure.
It needs to have both a "Live" and corresponding "edited" trace on the plot so changes can be made through mouse interaction without immediately pushing updates to the original. Having a local PV copy the original's contents means that it can be edited and manipulated as much as necessary before the live values are updated.

Is it possible to dynamically create local PVs based on the traces provided for a Plot (in/at start of runtime), or would each plotted PV/trace we want to be editable have to have a local copy defined in the .bob file (perhaps generated when adding a new widget for this functionality)?
Either way, ideally I'd be looking for some way to make this mostly automatic, besides flagging the traces/graph as editable in the first place.

Could anyone give some pointers on this? I'm still pretty new to Phoebus so anything would help really.
Thanks!

**Comment by georgweiss** [comment]
No sure this answers your question, but a local PV may be of array type. For instance, if a XY Plot widget is configured like so:
![Screenshot 2024-12-17 at 13 32 47](https://github.com/user-attachments/assets/c252feea-b7f5-4415-8cf4-cb6eb0458df4)
you would see this when running it:
![Screenshot 2024-12-17 at 13 32 54](https://github.com/user-attachments/assets/ec69f470-8364-4d08-820b-edde9bde8171)

**Comment by JonFijalkowski** [comment]
Hi George, thanks for the response.
That's good that I can store array values in local PVs, but I'm mainly seeing if local PVs be dynamically populated/created using the values from a Live PV (which for a plot widget, would be an array).
I suppose they could be created beforehand, only containing placeholder data, then replacing it when the live PVs connect?

I've got some other ideas so it may not be necessary to use Local PVs, but I'm expecting local PVs would be more responsive for things like mouse interaction, so would be favourable if it's possible to implement this way.

**Reply by georgweiss** [reply]
Are you looking for something like ```loc://x('pva://myPv')```?

**Reply by JonFijalkowski** [reply]
Oh neat, would that actually give a local copy? not just a local PV storing the name of the live PV?
If so that's amazingly straightforward and I wish I'd noticed it sooner hahah

**Reply by georgweiss** [reply]
Hold your horses, I did not mean to say that this is supported, because it isn't. Sadly. I asked to better understand your use case.

**Reply by JonFijalkowski** [reply]
Ah fair enough, seemed too good to be true.
As a general idea though, that's what I'm after. I could then generate that local PV automatically, or simply require something in that format as an additional editable trace.

**Reply by georgweiss** [reply]
Agree it would be useful to be able to initiate a local PV from the value of another PV. Question is how to handle error cases, e.g. if referenced PV is offline: one would not even be able to determine the type.

**Reply by JonFijalkowski** [reply]
Yeah, it'd be tricky for the system to know how to handle it, even if it's online - you can't determine anything automatically until the PV connects and you get a value.

I've got another idea or two to try for this part of my project, so I'm not stuck without options if stuff like this isn't possible.
I could add another Live PV (ignored/not used by the IOC) for storing the set of changes made so far, while using local PVs to track just the current coordinate changes being made. (if this makes any reasonable impact on performance, otherwise just using the secondary live PV for everything)
